### PR TITLE
Add a new deployer command to generate grafana tokens

### PR DESCRIPTION
**Pull request by GeorgianaElena**
This PR creates a new `deployer` command that generates (or re-generates) a Grafana API token and stores it into `enc-grafana-token.secret.yaml` file.

It also generates new tokens  (using this new cmd option), for the clusters in #1838 that have expired tokens.

Notable things that changed:
   - deployer grafana-related commands were moved into their own dir and module.
   - a few grafana-related functions from the `central_grafana.py` file were moved into an `utils` file
   -  the new command also allows to overwrite an existing token and shows whether it has expired or not
   - we are not generating API keys anymore, but tokens for service accounts as per Grafana's latest update 
      Screenshot 2022-11-24 at 14 52 27

More about Grafana service accounts here https://grafana.com/docs/grafana/latest/administration/service-accounts/ 

Closes https://github.com/2i2c-org/infrastructure/issues/1838
Reference https://github.com/2i2c-org/infrastructure/issues/1917

### CODE REVIEW

File: deployer/grafana/central_grafana.py
Comment (GeorgianaElena) [inline deployer/grafana/central_grafana.py:5]: This is not new content! It's the same code, with only some of the comments updates.
I believe this is why GitHub got fooled to believe it's a new file and not one that got moved 🙄

File: deployer/grafana/grafana_utils.py
Comment (GeorgianaElena) [inline deployer/grafana/grafana_utils.py:121]: Only this function is new. All the others used to exist in `central_grafana.py`.
There are only minor tweaks to the existing ones, like updating the function decription and fixing a small bug https://github.com/2i2c-org/infrastructure/pull/1938/commits/fe182d700f9533cf1974531f81b9e31944cb70c3

File: deployer/README.md
Comment (consideRatio) [inline deployer/README.md:213]: I love these kinds of summaries!!

Could you add some clarity what kinds of API tokens (a Grafana API token?), and service accounts (a grafana service account? a k8s service account?) are meant, and also what kind of credential is used to set things up?

Comment (GeorgianaElena) [reply deployer/README.md]: Added! Thanks @consideRatio! I spent this time with my head inside Grafana-stuff and forgot there exist other things  😅

File: deployer/grafana/grafana_tokens.py
Comment (yuvipanda) [inline deployer/grafana/grafana_tokens.py:34]: You can use https://github.com/2i2c-org/infrastructure/blob/348a88b436edcb14d104577a8479d9edd9a5ca84/deployer/file_acquisition.py#L72 to read the encrypted file, and it'll handle decryption correctly. We should also definitely use yaml to read the decrypted content, rather than using string parsing I think.

File: deployer/grafana/grafana_tokens.py
Comment (yuvipanda) [inline deployer/grafana/grafana_tokens.py:189]: Is there a reason we would want to keep an expired token? Could we automatically replace them if it exists and is expired?

**Review by yuvipanda — CHANGES_REQUESTED** [review]

File: deployer/grafana/grafana_tokens.py
Comment (GeorgianaElena) [reply deployer/grafana/grafana_tokens.py]: Haha, yes! I have used this before. I have no idea why this time my impulse was to reinvent the wheel do everything the ugly manual way 😅 Thanks for raising this

File: deployer/grafana/grafana_tokens.py
Comment (GeorgianaElena) [reply deployer/grafana/grafana_tokens.py]: Yes, no reason to keep them.
This script creates tokens with no expiration date however. I couldn't find in their docs how to set one https://grafana.com/docs/grafana/latest/developers/http_api/serviceaccount

**Comment by GeorgianaElena** [comment]
@2i2c-org/tech-team, signal boosting that this has addressed prior feedback and should now be ready for review again. Thank you!

File: deployer/README.md
Comment (sgibson91) [inline deployer/README.md:46]: Left over from a merge conflict?

File: deployer/README.md
Comment (sgibson91) [inline deployer/README.md:52]: merge conflict?

File: deployer/README.md
Comment (sgibson91) [inline deployer/README.md:208]:
```suggestion
This token will be used by the [`deploy-grafana-dashboards` workflow](https://github.com/2i2c-org/infrastructure/tree/HEAD/.github/workflows/deploy-grafana-dashboards.yaml) to authenticate with Grafana’s REST API
```

I mean, probably :D

File: deployer/grafana/central_grafana.py
Comment (sgibson91) [reply deployer/grafana/central_grafana.py:5]: Yeah the comment updates will always make git thinks it's a new file in diffs, unfortunately. From the object database point of view, this is a completely new object, and that's why it behaves in this way.

File: deployer/grafana/grafana_tokens.py
Comment (sgibson91) [inline deployer/grafana/grafana_tokens.py:81]: We already have 17 clusters - 18 once #2022 is merged. Assuming one grafana instance per cluster and they will all be using a service account, maybe best to bump this number now?

File: docs/hub-deployment-guide/deploy-support/setup-grafana-dashboards.md
Comment (sgibson91) [inline docs/hub-deployment-guide/deploy-support/setup-grafana-dashboards.md:35]:
```suggestion
If the command succeeded, it should have created:
```

small typo

**Review by sgibson91 — APPROVED** [review]
LGTM!

I have a small wondering on file naming. For example, now that we have a `grafana` subfolder, I would say that `grafana/utils.py` and `grafana/manage_tokens.py` would reduce redundancy compared to `grafana/grafana_utils.py` and `grafana/grafana_tokens.py` etc, particularly when importing, i.e. `from grafana import utils` versus `from grafana import grafana_utils`. In the `utils`-specific case we'd need to avoid naming clashes with the top-level `utils` though. Which indicates to me either (i) we are overusing `utils` and should be more creative in our filenaming, or (ii) I am overthinking package design :D

I don't think this should block merging this PR, but something for us to think about going forward.

File: deployer/grafana/grafana_tokens.py
Comment (sgibson91) [reply deployer/grafana/grafana_tokens.py:81]: Hmmm, I guess there are some long-lived tokens that will mean we might not get to this number?

File: deployer/README.md
Comment (GeorgianaElena) [reply deployer/README.md:208]: Grafana calls it HTTP API in their docs https://grafana.com/docs/grafana/latest/developers/http_api/ and I think this is why I called it http api also?

But I went and searched if there are any differences and found out that http api it's is own thing apparently https://www.programsbuzz.com/article/what-difference-between-rest-api-and-http-api.

I will link these in the docs to have them at hand. Thanks for sparking a dive into this topic. I had no idea they are different

Comment (sgibson91) [reply deployer/README.md]: Me either - that's super interesting!

File: deployer/grafana/grafana_tokens.py
Comment (GeorgianaElena) [reply deployer/grafana/grafana_tokens.py]: *` # Assumes we don't have more than 10 service accounts`* should have been *` # Assumes we don't have more than 10 service accounts per Grafana`*.

Sorry for the confusion. I added a commit to improve the comments around the service accounts per Grafana https://github.com/2i2c-org/infrastructure/pull/1938/commits/7645df1e66a1ba32a0bca6e82d3f8e379407700b.
 LMK is the comments are more clear now.

Basically, with this PR, we are creating the infrastructure to create our first service accounts **per Grafana**, called `deployer`. In case we will create other service accounts in the future for rbac purposes for example, and that number is >10 we should update this. But my guess is that this might take a while, if ever.

Comment (sgibson91) [reply deployer/grafana/grafana_tokens.py]: Ahhh, per grafana makes way more sense - thanks!

**Comment by GeorgianaElena** [comment]
> I have a small wondering on file naming. For example, now that we have a grafana subfolder, I would say that grafana/utils.py and grafana/manage_tokens.py would reduce redundancy compared to grafana/grafana_utils.py and grafana/grafana_tokens.py etc, particularly when importing, i.e. from grafana import utils versus from grafana import grafana_utils. In the utils-specific case we'd need to avoid naming clashes with the top-level utils though. Which indicates to me either (i) we are overusing utils and should be more creative in our filenaming, or (ii) I am overthinking package design :D

Very good point!! I am not proud at all about the naming of this files and to be honest, I have always been bad at naming! Do you mind if I open an issue about improving the design after merging this PR? I just notice the same naming pattern for the `deployer/generate` directory too (also my doing 🤦🏼‍♀️).

**Comment by sgibson91** [comment]
@GeorgianaElena a follow up issue is perfect, thank you! And I agree, naming things is hard :)

**Pull request merged**
